49 unpublish events

Fabian-Sommer opened the pull request:
(#49)

Fabian-Sommer committed:
Added published field to database

Fabian-Sommer committed:
Added label for organizers. Drafts are not shown for pupils (non-organizers/admins).
Removed old drafts button and draft field in db. Changed tests accordingly.

Baschdl committed:
Merge branch 'dev' into 49_unpublish_events

Baschdl committed:
Merge branch 'dev' into 49_unpublish_events

tom95 commented on app/models/event.rb:
```diff
@@ -222,7 +222,7 @@ def material_path
   end
 
-  scope :draft_is, ->(draft) { where("draft = ?", draft) }
+  scope :draft_is, ->(published) { where("published = ?", published) }
```
You should rename this or negate the published

tom95 commented on app/models/event.rb:
```diff
@@ -231,7 +231,7 @@ def material_path
   # @return List of events
   def self.sorted_by_start_date(only_public)
-    (only_public ? Event.draft_is(false) : Event.all)
+    (only_public ? Event.draft_is(true) : Event.all)
```
Change from above should be reflected here, so that it makes sense again

tom95 commented on app/views/events/_event.html.erb:
```diff
@@ -1,3 +1,4 @@
+<% if event.published or can? :view_unpublished, Event %>
```
I would say this should be filtered out in the controller, starting from the query, and not in the view.

tom95 commented on app/views/events/_event.html.erb:
```diff
@@ -11,7 +12,8 @@
     <small>
-      <%= tag_label t('activerecord.attributes.event.draft'), :default if event.draft %>
+      <%= tag_label t('events.notices.is_published'), :success if event.published and can? :view_unpublished, Event %>
+      <%= tag_label t('activerecord.attributes.event.draft'), :default if not event.published and can? :view_unpublished, Event %>
```
I think the requirement to show unpublished/published on the view came from the old table layout. In the new layout, just showing a draft label should be enough, since any other event is of course published. @hpi-swt2/product-owner-2016

Fabian-Sommer replied:
This request was added to US 3 days ago (at least to title of US)

tom95 commented on db/migrate/20161126190339_add_draft_to_events.rb:
```diff
@@ -1,5 +0,0 @@
-  def change
-    add_column :events, :draft, :boolean
-  end
-end
```
You won't want to remove any migrations ever. It also seems you forgot to add your own migration that would add the published field.

bjrne replied:
[This page](http://edgeguides.rubyonrails.org/active_record_migrations.html#running-migrations) has everything you need to know about migrations. I really recommend to read it through.

tom95 reviewed (changes requested):
I don't see why you had to rename the `draft` field to `published`. This change really cluttered the merge request and introduced a couple of mistakes in this MR. Was there a need to rename the field?

Fabian-Sommer committed:
Requested changes

Fabian-Sommer committed:
Merge branch 'dev' into 49_unpublish_events

Conflicts:
	app/models/ability.rb
	db/schema.rb

aloeser commented on db/migrate/20170111135212_rename_draft_to_published.rb:
```diff
@@ -0,0 +1,5 @@
+class RenameDraftToPublished < ActiveRecord::Migration
+  def change
+    rename_column :events, :draft, :published
```
Dont you need to invert the values too? Otherwise the unpublished (drafts) will become published, and the other way round

Fabian-Sommer replied:
Is that a problem (I repopulate sample data when pulling from dev)?

bjrne replied:
It isn't now, but if our System would be deployed already, it would be.

aloeser replied:
Repopulate sample data sounds for me like a database reset. I dont think you want to reset the database with every single commit/merge when running in production mode.

bjrne commented on app/controllers/events_controller.rb:
```diff
@@ -7,7 +7,7 @@ class EventsController < ApplicationController
   # GET /events
   def index
-    @events = Event.sorted_by_start_date(!can?(:edit, Event)).reverse
+    @events = Event.sorted_by_start_date(!can?(:view_unpublished, Event))
```
Is there a reason you removed the `.reverse`?

aloeser replied:
btw, you could use `cannot?` instead of `!can?`

Fabian-Sommer replied:
I dont think it is good to have the latest (and therefore probably most uninteresting) events shown first and events that are about to begin (and therefore interesting) towards the bottom. Also, this achieves consistency with the start page. 
Is there a reason for showing them in different orders on start page and event page @tom95 @hpi-swt2/product-owner-2016 ?

bjrne replied:
I was just wondering, because this is part of our user story #304 and I was afraid we would loose track if things are changed on the fly in PRs that have little to do with it.

Fabian-Sommer replied:
Since no one is working on #304 yet and it is not really related to an archive page anyway, I suggest just merging it here and deleting it from #304.

tom95 replied:
Yes I think we'll tackle ordering as part of #304 once this is merged. Original intent was, since the main page starts in the present, to show the next event (so ascending order) and on the list page use descending order, since we don't want to start looking at all the past events.

Fabian-Sommer committed:
Reversing values of draft when migrating

Fabian-Sommer committed:
Removed "is published" label

Fabian-Sommer committed:
Merge branch 'dev' into 49_unpublish_events

tom95 commented on app/views/events/_form.html.erb:
```diff
@@ -88,16 +88,27 @@
       <% else %>
-        <%= f.submit t('.publish'), :name => 'publish', :class => 'btn btn-primary' %>
-        <%= f.submit t('.draft'), :name => 'draft', :class => 'btn btn-default' %>
+        <%= f.submit t('.create'), :name => 'create', :class => 'btn btn-primary' %>
```
The label "create" currently appear whenever you have an unpublished event, which could be confusing. I think the `if` above should just ask for `@event.new_record?` to determine whether to show create or update.

tom95 commented on app/assets/stylesheets/participants.css:
```diff
@@ -0,0 +1,3 @@
+#participants_download_submit_button  {
+    float: right;
+}
```
I would guess these changes sneaked in through a different branch?

Fabian-Sommer replied:
This should have been part of PR #259, but I forgot to push it

tom95 reviewed (changes requested):
Thanks for adapting everything! After those two small fixes I'd say we can finally merge this branch.

Fabian-Sommer committed:
Changed create button to update button for unpublished existing events

tom95 committed:
Merge branch 'dev' into 49_unpublish_events

tom95 committed:
Add debug output to check why the test is failing on travis

tom95 committed:
oops

tom95 committed:
oops oops

tom95 committed:
Use activerecord queries instead of sql

aloeser commented on db/migrate/20170111135212_rename_draft_to_published.rb:
```diff
@@ -0,0 +1,14 @@
+        Event.all.each {|e| e.toggle!(:draft)}
+      end
+      dir.down do
+        Event.all.each {|e| e.toggle!(:published)}
```
I think it should be :draft here too. When i run the migration it works, but when I try a rollback it fails with the error that Event has no published attribute. I guess the rename happens first

aloeser committed:
Update 20170111135212_rename_draft_to_published.rb

When rolling back this migration, toggle the attribute "draft", not "published"
